refactor(error): improve error handling

dan-da opened the pull request:
Addresses issue #247.  (part a: more to come later)

I started small by just fixing 4 unwrap() calls in kindelia_core/src/net.rs.

Unfortunately one of these involved a trait, which complicated things a bit and required small changes in other files.

In kindelia `main.rs` I could have just mapped the errors from net.rs into a string and submitted a smaller PR, but instead I decided it would be cleaner to handle them as `anyhow::Result`.

I ended up changing most fn in `main.rs` and `config.rs` to return `anyhow::Result` instead of `Result<_, String>`.   This required some calls to `map_err(|e| anyhow!(e)` to deal with `String` errors from `format!()` or from lower level fn calls.  However, as the lower level calls get fixed, then the `map_err` goes away and we will just have `fn_call()?`.   very clean.

note: the map_err() is necessary because `anyhow::Result` accepts `anyhow::Error` which is a boxed dynamic error type that accepts any error type derived from `std::error::Error`.   However `String` is not so derived and thus must be mapped (or preferably return `anyhow::Result` or a thiserror type instead).   

So this is very much a preliminary / intermediate PR to introduce anyhow and show the direction.

I understand there is a big refactor happening in another branch.   I am ok with rebasing onto it when merged.   Or this could be merged first, whichever works.

------

kindelia_core:

* removes all unwrap() from net.rs
* defines error enums in net.rs: ProtoCommError and ParseAddressError
* return Result from affected fn in net.rs
* change IPV6 panic!() to unimplemented!()
* accept addr arg for Node::new() so it can avoid returning Result
* update test and bench as necessary

kindelia:

* add dep on anyhow crate
* return anyhow::Result from config methods
* return anyhow::Result from main methods, including main()
* return anyhow::Result from FileInput::read_to_string()

dan-da commented:
Something seems wonky with CI.   afaict, checks are only running when the PR is first created, and not for subsequent push(es).    or at least it does not run for *some* subsequent pushes.    is this intentional?

dan-da commented:
ok, so I was getting some emails from github that CI was failing.  Here's an example:
https://github.com/dan-da/Kindelia/actions/runs/3562238811

It was some kind of cargo bench issue.  I found a Cargo.toml workaround, and pushed changes for it to this branch.  After which, I get another kind of failure:
https://github.com/dan-da/Kindelia/actions/runs/3562393608

```
Run racs4/github-action-benchmark@v1
/usr/bin/git -c user.name=github-action-benchmark -c user.email=github@users.noreply.github.com -c http.https://github.com/.extraheader= fetch ***github.com/dan-da/Kindelia.git gh-pages:gh-pages
fatal: couldn't find remote ref gh-pages
Error: Command 'git' failed with args '-c user.name=github-action-benchmark -c user.email=github@users.noreply.github.com -c http.https://github.com/.extraheader= fetch ***github.com/dan-da/Kindelia.git gh-pages:gh-pages': fatal: couldn't find remote ref gh-pages
: Error: The process '/usr/bin/git' failed with exit code 128
```

maybe it wants me to have a gh-pages branch for some reason?

racs4 commented:
> Something seems wonky with CI

I think this is solved in ab32905, but you can talk if you have any problems.

> then the map_err goes away and we will just have fn_call()?. very clean

This seems very good.

> Or this could be merged first

I think this is better, since this is smaller.

racs4 commented on kindelia_core/src/net.rs:
```diff
@@ -72,23 +87,46 @@ impl std::fmt::Display for Address {
-  let vals =
-    strs[0].split('.').map(|o| o.parse::<u8>().unwrap()).collect::<Vec<u8>>();
-  let port = strs.get(1).map(|s| s.parse::<u16>().unwrap()).unwrap_or(UDP_PORT);
-  Address::IPv4 {
+  let vals = strs[0]
+    .split('.')
+    .map(|o| {
+      o.parse::<u8>()
+        .map_err(|e| ParseAddressError::InvalidAddress{value: strs[0].to_string(), source: e}) })
+    .collect::<Result<Vec<u8>, ParseAddressError>>()?;
+  let port = match strs.get(1) {
+    Some(s) => s
+      .parse::<u16>()
+      .map_err(|e| ParseAddressError::InvalidAddress{value: strs[0].to_string(), source: e})?,
+    None => UDP_PORT,
+  };
+  Ok(Address::IPv4 {
     val0: vals[0],
     val1: vals[1],
     val2: vals[2],
     val3: vals[3],
     port,
-  }
+  })
 }
```
These `array[n]` type accesses can also cause unwanted `panics`. Doing this code without them however would be pretty boring.

I don't know if there was a reason why this address wasn't created as a `std::net::SocketAddrV4` or a `std::net::SocketAddr` (the latter is an enum that already contains IPv6 socket addresses, which would resolve existing IPv6 TODOs in the code); both contain a `.parse()` method which already does this parsing work for us. I'll talk to @steinerkelvin about it and I think I can change it, but it will greatly increase the scope of this PR if done here (many changes to `node.rs` will be needed).

Maybe we can leave it as it is in this PR, and open a new issue for this to be done.

dan-da replied:
> These array[n] type accesses can also cause unwanted panics. Doing this code without them however would be pretty boring.

great point.  yeah, I think we could deal with these in another pass.  I will add it to my list.   Shouldn't be too bad, something like: 

```
vals.get(0).ok_or(ArrayOutofBounds(0))?
```

which could even become a macro:  `aget!(vals,0)?` or maybe use a trait `vals.eget(0)?`.  Maybe add some context.   Likely there is already a crate for this if we look.

btw, I found there is a crate [no_panic](https://docs.rs/no-panic/latest/no_panic/) that provides a macro which forces compiler to verify that a function *cannot* panic and provides that guarantee to callers.  Unfortunately many std lib calls panic, so it can't be used everywhere.  I'll play around with it also.  I'm not certain yet how much value it can provide us.

> std::net::SocketAddr

yeah, I've been thinking it would be best to use that, or possibly find a crate with a higher level type that accepts either domain:port or ip:port (v4 or v6) and resolves domains securely, at least for accepting cli/config args.   

> Maybe we can leave it as it is in this PR, and open a new issue for this to be done

yeah, definitely something for a separate PR.  outside the scope here.

dan-da commented on kindelia_core/src/node.rs:
```diff
@@ -760,9 +760,11 @@ pub fn miner_loop(
   pub fn new(
     data_path: PathBuf,
     network_id: u32,
+    addr: C::Address,
```
happy to add a comment, but I'm not yet seeing why `addr` should be removed...?

btw, there is presently only one `expect()` in `Node::new()` and it is removed in #250.

When possible, I prefer that `::new()` methods accept all arguments in a *final* state so that calls that can panic are unnecessary within `::new()`.  Just seems cleaner for the API.  So hopefully we can achieve that here.   Sometimes builder pattern is helpful. 

btw, for this PR, the focus is really just on `kindelia_core/net.rs` and `kindelia/main.rs`.  Any changes to other files are just to make those work.  I have other (mostly smaller) PRs coming that focus on eg node.rs, persistence.rs, etc, one at a time.

racs4 replied:
> happy to add a comment (just did), but I'm not yet seeing why addr should be removed...?

Well, I commented this thinking that the `addr: C::Address` parameter will always be related to the `comm: C` parameter, so that whenever you want to use a function that receives both, you would always have to do the same thing and use `comm.get_addr()`.

But in addition, with these parameters separated, the function has the possibility of receiving an address different from the address in `comm`; considering the implementation of `UdpSocket` one could create a struct of this type by binding it to an X address, but passing a Y address to `Node::new`, which would cause confusion in the node, making it having a real address and a nominal address (of course we can simply check if this inconsistency exists the few times the `Node::new` function is called, I'm just saying that there would be that possibility).

dan-da replied:
ok, yeah totally makes sense.  I forgot about comm.get_addr().

dan-da commented:
I just force pushed with these changes:
1. removed Cargo.toml bench sections to disable benches
2. added a comment to Node::new() as [requested](https://github.com/Kindelia/Kindelia-Chain/pull/252#discussion_r1037732536)
